test/unit/test_stream_to.cxx

```diff
@@ -373,6 +373,49 @@ void test_stream_to()
   test_variant_fold(conn);
 }
 
+void test_nonoptionals_binary(pqxx::connection &connection)
+{
+  pqxx::work tx{connection};
+  auto inserter{pqxx::stream_to::table(tx, {"stream_binary_to_test"}, {}, true)};
+  PQXX_CHECK(inserter, "stream_to (binary) failed to initialize");
+
+  auto const nonascii{"\u3053\u3093\u306b\u3061\u308f"};
+  bytea const binary{'\x00', '\x01', '\x02'},
+    text{'f', 'o', 'o', ' ', 'b', 'a', 'r', '\0'};
+
+  inserter << std::make_tuple(
+    1234, binary);
+  inserter << std::make_tuple(
+    5678, text);
+  inserter << std::make_tuple(910, bytea{});
+
+  inserter.complete();
+
+  auto r1{tx.exec1("SELECT * FROM stream_binary_to_test WHERE number0 = 1234")};
+  PQXX_CHECK_EQUAL(r1[0].as<int>(), 1234, "Read back wrong first int.");
```

Comment:
How is this going to work?  AIUI the binary format has a very different, more complicated number format.  You can't just take the address of an `int` and send it across.

(This is part of the bigger problem that I brought up in the beginning: the binary formats are complex, undocumented, and not necessarily stable.)

Also, I see nothing to deal with byte orders.  IIRC the binary format generally works in Little-Endian mode — what if the client system runs in Big-Endian mode?

And finally how do nulls work?

Reply:
Test is incomplete (and not correct), yes.
htonl[s]/ntoh[l|s]  etc. must be used.

For nulls:
https://www.postgresql.org/docs/16/sql-copy.html#:~:text=As%20a%20special%20case%2C%20%2D1%20indicates%20a%20NULL%20field%20value.%20No%20value%20bytes%20follow%20in%20the%20NULL%20case.

But all in all would be better to provide a wrapper around the processing of the header, tuples and trailer, and provide a consistent api here.

for example the when reading data, open method would go something like this:
  use get raw_line from stream
  check signature
  read header
      make some assumptions on the header flag values (or ignore them, depending on server version)
      get number of fields

Then get data tuples from stream and continue or get more data, read field sizes, ensure it's consistent with the provided types client specified, read data and convert from network format.

Reply:
By the way, as of C++20 we're getting [at least _some_ support](https://en.cppreference.com/w/cpp/types/endian) for byte-swapping.  (Although it doesn't look [complete](https://en.cppreference.com/w/cpp/numeric/byteswap) until C++23.)

My impression is that things get _really_ hard where it comes to getting the SQL-side types and the C++-side types matched up correctly.  For example, if the database is sending an integer but we're trying to read it as a floating-point number, we'll still need to start by decoding the binary value as the right type — and then maybe we'd then want to convert the integer to a floating-point value after that, or maybe we shouldn't even allow that to avoid mistakes.  So it's possible that we'll need an entirely different paradigm for working with the types.